### PR TITLE
adding full humann2 pipeline

### PR DESCRIPTION
**Pull request by antgonza**
This adds the full humann2 pipeline. However, currently I'm testing that it fails because I'm seeing some strange behavior with `humann2_join_tables` as it's not generating the expected output (and it doesn't raise an error) so it fails in the next step as the files are not generated. I have contacted the developers to check what's going on.

Anyway, I think it will be good to review (@josenavas, @tanaes) so we can discuss some of this points and perhaps merge so @tanaes can continue development of kneaddata.

### CODE REVIEW

File: qp_shotgun/humann2/humann2.py [hunk 1]
```diff
@@ -137,21 +159,45 @@ def humann2(qclient, job_id, parameters, out_dir):
...
+    commands.append(('humann2_join_tables -i {0} -o {0}/pathabundance.biom '
+                    '--file_name pathabundance_relab').format(out_dir))
+    msg = "Step 4 of 5: Merging resulting tables job (%d/3)"
+    success, msg = _run_commands(qclient, job_id, commands, msg)
+    if not success:
+        return False, None, msg
+
+    # Step 5 generating re-normalized tables
+    commands = []
+    commands.append(('humann2_renorm_table -i {0}/genefamilies.biom -u cpm '
```
Comment (tanaes) [inline qp_shotgun/humann2/humann2.py:183]: Why are we normalizing the genefamilies tables to cpm but others as relabund?

File: qp_shotgun/humann2/humann2.py [hunk 2]
```diff
@@ -137,21 +159,45 @@ def humann2(qclient, job_id, parameters, out_dir):
...
     # Step 4 merge tables
-    qclient.update_job_step(job_id, "Step 4 of 5: Merging resulting tables")
-
-    # Step 5 generating re-normalized tables: TODO
-    qclient.update_job_step(job_id, "Step 5 of 5: Re-normalizing tables")
-    artifacts_info = []
-
-    return True, artifacts_info, ""
+    commands = []
+    commands.append(('humann2_join_tables -i {0} -o {0}/genefamilies.biom '
```
Comment (tanaes) [inline qp_shotgun/humann2/humann2.py:167]: Note that these are going to be by default the funny stratified tables, so that there will be a line for the gene family and then a bunch of lines for that same gene family split across each of the different genomes that had hits contributing to it. (in other words, each annotated read gets represented twice per table)

We don't have a lot of tools to deal with this data structure right now -- maybe want to include the --remove-stratified-output flag to have simple count tables for now.

**Review by tanaes — APPROVED** [review]

File: qp_shotgun/humann2/humann2.py (hunk 2 above)
Comment (antgonza) [reply qp_shotgun/humann2/humann2.py:167]: Thanks, that's one of the things we are working with the devs.

File: qp_shotgun/humann2/humann2.py (hunk 1 above)
Comment (antgonza) [reply qp_shotgun/humann2/humann2.py:183]: That's the suggestion of the developers.

Comment (tanaes) [reply qp_shotgun/humann2/humann2.py:183]: 👍

File: qp_shotgun/humann2/humann2.py [hunk 3]
```diff
@@ -91,7 +91,8 @@ def generate_humann2_analysis_commands(forward_seqs, reverse_seqs, map_file,
             % ', '.join(sn_by_rp.keys()))
 
     cmds = []
-    params = ['--%s "%s"' % (k, v) for k, v in viewitems(parameters) if v]
+    params = ['--%s "%s"' % (k, v) if v is not True else '--%s' % k
+              for k, v in viewitems(parameters) if v]
```
Comment (tanaes) [inline qp_shotgun/humann2/humann2.py:93]: Note that I've had problems with the integer 1 here in this context...

Here's my current workaround:

https://github.com/tanaes/qp-shotgun/blob/adding-kneaddata-pipeline/qp_shotgun/kneaddata/kneaddata.py#L103

Comment (antgonza) [reply qp_shotgun/humann2/humann2.py:93]: Interesting, I don't know why it didn't work for you but this works for me:

``` python
In [2]: a = [True, 1, 'text', False]

In [4]: [v for v in a if v]
Out[4]: [True, 1, 'text']

In [5]: [v if v is not True else 'This is True!!' for v in a if v]
Out[5]: ['This is True!!', 1, 'text']
```

File: README.rst [hunk 4]
```diff
@@ -23,6 +23,7 @@ The default pipeline is:
 2. Collate individual results into single OTU tables: gene families, path coverage and path abundance.
 3. Renormalize the tables: gene families - CPM, path coverage - relative abundance, path abundance - relative abundance.
 
+Note that HUMAnN2 depends on metaphlan2 but neither the HUMAnN2 or this plugin automatically install that dependency.
```
Comment (josenavas) [inline README.rst]: Probably useful to add a link to the metaphlan2 installation instructions?

File: qp_shotgun/humann2/__init__.py [hunk 5]
```diff
@@ -62,7 +62,8 @@
     'input-format': ['choice:["", "fastq", "fastq.gz", "fasta", "fasta.gz", '
                      '"sam", "bam", "blastm8", "genetable", "biom"]', ''],
     'pathways': ['choice:["metacyc", "unipathway"]', 'metacyc'],
-    'memory-use': ['choice:["minimum", "maximum"]', 'minimum']}
+    'memory-use': ['choice:["minimum", "maximum"]', 'minimum'],
+    'remove-column-description-output': ['boolean', 'True']}
```
Comment (josenavas) [inline qp_shotgun/humann2/__init__.py:66]: If we pass false to this parameter it will break the pipeline right? In that case I don't think we should expose this parameter.

File: qp_shotgun/humann2/humann2.py [hunk 6]
```diff
@@ -86,15 +91,32 @@ def generate_humann2_analysis_commands(forward_seqs, reverse_seqs, map_file,
...
     for ffn, fn, s in samples:
+        od = join(out_dir, fn)
+        # just making sure the output directory exists
+        if assure_od and not exists(od):
```
Comment (josenavas) [inline qp_shotgun/humann2/humann2.py]: Why is this needed? It looks like this is just used to cheat on the tests.

File: qp_shotgun/humann2/humann2.py [hunk 7]
```diff
@@ -137,21 +159,48 @@ def humann2(qclient, job_id, parameters, out_dir):
...
+
+    # Generating artifact
+    pb = partial(join, out_dir)
+    filepaths = [(pb('genefamilies.biom'), 'biom'),
+                 (pb('pathcoverage.biom'), 'biom'),
+                 (pb('pathabundance.biom'), 'biom'),
+                 (pb('genefamilies_cpm.biom'), 'biom'),
+                 (pb('pathcoverage_relab.biom'), 'biom'),
+                 (pb('pathabundance_relab.biom'), 'biom')]
+    ainfo = [ArtifactInfo('OTU table', 'BIOM', filepaths)]
```
Comment (josenavas) [inline qp_shotgun/humann2/humann2.py]: I don't think the output from humann2 is an OTU table. And you've actually defined the output of HUMANn2 to be per sample FASTQ [here](https://github.com/qiita-spots/qp-shotgun/pull/7/files#diff-2653c686b56da473154466db0a8a5b85R67)

File: setup.py [hunk 8]
```diff
@@ -41,7 +43,7 @@
       test_suite='nose.collector',
       packages=['qp_shotgun', 'qp_shotgun/humann2'],
       package_data={'qp_shotgun': ['support_files/config_file.cfg']},
-      scripts=glob('scripts/*'),
+      scripts=['scripts/configure_shotgun', 'scripts/start_shotgun'],
```
Comment (josenavas) [inline setup.py:46]: Just curiosity - why this change?

**Review by josenavas — CHANGES_REQUESTED** [review]
Just a few changes and a question

File: README.rst (hunk 4 above)
Comment (antgonza) [reply README.rst]: Good idea.

Comment (antgonza) [reply README.rst]: I'm also adding some basic instuctions.

File: qp_shotgun/humann2/__init__.py (hunk 5 above)
Comment (antgonza) [reply qp_shotgun/humann2/__init__.py:66]: Well, it will not break the pipeline but we won't be able to generate plots using the current metadata as it is stored. I'm OK with leaving it as it is but if you feel really strong about it, just let me know.

File: qp_shotgun/humann2/humann2.py (hunk 6 above)
Comment (antgonza) [reply qp_shotgun/humann2/humann2.py]: That's correct and it's already mentioned in the docstring, any suggestions?

File: qp_shotgun/humann2/humann2.py (hunk 7 above)
Comment (antgonza) [reply qp_shotgun/humann2/humann2.py]: Funny enough, it is 3 biom tables. I will change in the link you sent, is there anything else I need to do?

File: setup.py (hunk 8 above)
Comment (antgonza) [reply setup.py:46]: Part of the long set of tests to install metaphlan2.py, nothing in particular ...

File: qp_shotgun/humann2/humann2.py (hunk 7 above)
Comment (josenavas) [reply qp_shotgun/humann2/humann2.py]: If I understand correctly, the output is 3 artifacts, each of them have 2 biom tables, right?
- Gene families biom and normalized biom
- Path coverage biom and normalized biom
- Path abundance biom and normalized biom

If that's correct, I would create 3 different artifact info object and expose that this command has 3 outputs.

An alternative would be to describe 6 artifacts and then we "decorate" the artifact types of 3 of them to be normalized. Does it make sense? However, the decoration is still not implemented (I'm working on it) so we can change later if you want.

Comment (josenavas) [reply qp_shotgun/humann2/humann2.py]: Thinking more about it (it just hit me now xD) I think is better to create 6 artifacts. The reason why is because in the analysis pipeline, you are expecting a biom table, so you want to be clear which biom do you want to use: normalized or not, and if it contains gene or path abundance/coverage

File: qp_shotgun/humann2/humann2.py (hunk 6 above)
Comment (josenavas) [reply qp_shotgun/humann2/humann2.py]: Is this a result of humann2 not being executed on travis? otherwise I don't understand why the tests are checking if the directory exists if this function is only creating the command.

i.e. if this is added to cheat on the tests, either there is something wrong on the tests, or there is something wrong on the code.

**Comment by jdereus** [comment]
Can one of the admins verify this patch?

**Comment by josenavas** [comment]
add to  whitelist

File: .travis.yml [hunk 9]
```diff
@@ -22,13 +22,18 @@ install:
   - qiita-env make --no-load-ontologies
   - export QIITA_SERVER_CERT=$PWD/miniconda3/envs/qiita_env/lib/python2.7/site-packages/qiita_core/support_files/server.crt
   - pip install https://github.com/qiita-spots/qiita_client/archive/master.zip
+  - wget https://bitbucket.org/biobakery/metaphlan2/get/default.zip
```
Comment (josenavas) [inline .travis.yml:30]: You should move these lines after the `pip install .` because at this point humann2 is still not installed

**Review by josenavas — CHANGES_REQUESTED** [review]
Just one comment so tests pass

File: qp_shotgun/humann2/humann2.py (hunk 2 above)
Comment (tanaes) [reply qp_shotgun/humann2/humann2.py:167]: What is the current behavior as of now? Are we removing the stratified output? Do you have an example output file I could take a look at?

File: qp_shotgun/humann2/humann2.py [hunk 10]
```diff
@@ -137,21 +157,54 @@ def humann2(qclient, job_id, parameters, out_dir):
...
+    commands.append(('humann2_join_tables -i {0} -o {0}/genefamilies.biom '
+                     '--file_name genefamilies --search-subdirectories '
+                     '--verbose').format(out_dir))
+    commands.append(('humann2_join_tables -i {0} -o {0}/pathcoverage.biom '
+                     '--file_name pathcoverage --search-subdirectories '
+                     '--verbose').format(out_dir))
+    commands.append(('humann2_join_tables -i {0} -o {0}/pathabundance.biom '
+                     '--file_name pathabundance --search-subdirectories '
+                     '--verbose').format(out_dir))
+    msg = "Step 4 of 5: Merging resulting tables job (%d/3)"
```
Comment (tanaes) [inline qp_shotgun/humann2/humann2.py:176]: I think it would be nice to also extract the purely taxonomic output from Metaphlan. 

Currently, this could be done by passing the `--metaphlan-options "--biom [biom filepath"` as one of the inputs.

These would need to be combined then as well.

File: qp_shotgun/humann2/tests/test_humann2.py [hunk 11]
```diff
@@ -132,8 +136,11 @@ def test_generate_humann2_analysis_commands_forward_reverse(self):
             f.write(MAPPING_FILE)
         self._clean_up_files.append(fp)
 
+        out_dir = mkdtemp()
```
Comment (tanaes) [inline qp_shotgun/humann2/tests/test_humann2.py:139]: Do we want to make this mapping file in setUp and include fp as a test class variable?

File: qp_shotgun/humann2/tests/test_humann2.py [hunk 12]
```diff
@@ -146,8 +153,8 @@ def test_generate_humann2_analysis_commands_forward_reverse(self):
             '--xipe "off" --annotation-gene-index "8" '
             '--protein-database "uniref" --threads "1" --pathways "metacyc" '
             '--pick-frames "off" --translated-alignment "diamond" '
-            '--log-level "DEBUG"',
-            'humann2 --input "fastq/s1.R2.fastq" --output "output/s1.R2" '
+            '--remove-column-description-output --log-level "DEBUG"' % out_dir,
```
Comment (tanaes) [inline qp_shotgun/humann2/tests/test_humann2.py:156]: Is the % out_dir on the wrong line here? should it go after `--output "%s/s1.R2" '` on line 157?

...and on line 143,172, 185, 200, 214, and 227?

**Comment by antgonza** [comment]
Some of @tanaes comments were done in a previous version so replying here:
- stratified output: This is the official reply from the devs: 
  `We have had a couple questions from users about how to break up the output files into stratified/unstratified files. We ended up adding a tool named "humann2_split_stratified_table". This will take the output files and split them into two files. The first file has the overall abundances and the second file has the abundance split by taxonomy. My recommendation would be to use this script at the end of your pipeline (ie after renorm). Is this something that sounds like it would work for you? If so, let me know and I can add biom support to the script (currently it only supports tsv files)`. Thus, decided to leave the default behavior, and users can run that script after.
-  mapping file in setUp: I think we could go that way but both options have pros/cons, if you feel strongly about it I can change.
- % out_dir: The current placement is fine. You can corroborate this because the tests are passing and flake8 is not complaining.

**Comment by antgonza** [comment]
After offline discussion we agreed that the easiest was to create an issue about point 1 (https://github.com/qiita-spots/qp-shotgun/issues/14) and continue development.

**Comment by tanaes** [comment]
Cool, merging PR
